pkg/controller/podautoscaler/replica_calculator.go

```diff
@@ -241,20 +241,30 @@ func (c *ReplicaCalculator) GetObjectMetricReplicas(currentReplicas int32, targe
 	}
 
 	usageRatio := float64(utilization) / float64(targetUtilization)
-	if math.Abs(1.0-usageRatio) <= c.tolerance {
-		// return the current replicas if the change would be too small
-		return currentReplicas, utilization, timestamp, nil
-	}
-
-	readyPodCount, err := c.getReadyPodsCount(namespace, selector)
+	replicaCount, timestamp, err = c.getUsageRatioReplicaCount(currentReplicas, usageRatio, namespace, selector)
+	return replicaCount, utilization, timestamp, err
+}
 
-	if err != nil {
-		return 0, 0, time.Time{}, fmt.Errorf("unable to calculate ready pods: %s", err)
+// getUsageRatioReplicaCount calculates the desired replica count based on usageRatio and ready pods count.
+// For currentReplicas=0 doesn't take into account ready pods count and tolerance to support scaling to zero pods.
+func (c *ReplicaCalculator) getUsageRatioReplicaCount(currentReplicas int32, usageRatio float64, namespace string, selector labels.Selector) (replicaCount int32, timestamp time.Time, err error) {
+	if currentReplicas != 0 {
+		if math.Abs(1.0-usageRatio) <= c.tolerance {
+			// return the current replicas if the change would be too small
+			return currentReplicas, timestamp, nil
+		}
+		readyPodCount := int64(0)
+		readyPodCount, err = c.getReadyPodsCount(namespace, selector)
+		if err != nil {
+			return 0, time.Time{}, fmt.Errorf("unable to calculate ready pods: %s", err)
+		}
+		replicaCount = int32(math.Ceil(usageRatio * float64(readyPodCount)))
+	} else {
+		// Scale to zero or n pods depending on usageRatio
+		replicaCount = int32(math.Ceil(usageRatio))
```

Comment:
This logic is not working in our case. We are scaling our deployment from min to max based on SQS queue metric. Suppose we want to scale the deployment `sqsworker` from `min=1` to `max=20` pods. Scaling should happen when the queue length is greater than 100.

With this logic when we scale down to 0. There is no way to come to 1 untill the queue length becomes > 100.

```
ApproximateNumberOfMessages = 0
    usageRatio = 0/100 = 0
    currentReplicas = 1
    readyPodCount = 1
        desired = int32(math.Ceil(usageRatio * float64(readyPodCount)))
                = int32(0*1) = 0

    ApproximateNumberOfMessages = 2
    usageRatio = 2/100 = 0.02
    currentReplicas = 0
    readyPodCount = 0
        desired = int32(math.Ceil(usageRatio * float64(readyPodCount)))
                = int32(ceil(0.02*0)) = 0
```